metagpt/rag/benchmark/base.py

```diff
@@ -0,0 +1,135 @@
+import evaluate
+import jieba
+import asyncio
+import os
+import json
+
+from typing import List
+from pydantic import BaseModel
+
+from llama_index.core.evaluation import SemanticSimilarityEvaluator
+from llama_index.core.embeddings import BaseEmbedding
+from llama_index.core.schema import NodeWithScore
+
+from metagpt.rag.factories import get_rag_embedding
+from metagpt.const import EXAMPLE_BENCHMARK_PATH
+
+
+class DatasetInfo(BaseModel):
+    name: str
+    document_files: List[str]
+    gt_info: List[dict]
+
+class DatasetConfig(BaseModel):
+    datasets: List[DatasetInfo]
+
+
+class RAGBenchMark:
+
+    def __init__(self,
+        embed_model: BaseEmbedding = None,
+         ):
+
+        self.evaluator = SemanticSimilarityEvaluator(
+            embed_model=embed_model or get_rag_embedding(),
+        )
+
+
+    def bleu_score(
+        self,
+        response: str,
+        reference: str,
+        with_penalty=False
+    ) -> float:
+        f = lambda text: list(jieba.cut(text))
+        bleu = evaluate.load(path = './metrics/bleu')
+        results = bleu.compute(predictions=[response], references=[[reference]], tokenizer=f)
+
+        bleu_avg = results['bleu']
+        bleu1 = results['precisions'][0]
+        bleu2 = results['precisions'][1]
+        bleu3 = results['precisions'][2]
+        bleu4 = results['precisions'][3]
+        brevity_penalty = results['brevity_penalty']
+
+        if with_penalty:
+            return bleu_avg, bleu1, bleu2, bleu3, bleu4
+        else:
+            return 0.0 if brevity_penalty == 0 else bleu_avg / brevity_penalty, bleu1, bleu2, bleu3, bleu4
+
+    def rougeL_score(
+        self,
+        response: str,
+        reference: str
+    ) -> float:
+        # pip install rouge_score
+        f = lambda text: list(jieba.cut(text))
+        rouge = evaluate.load(path = './metrics/rouge')
+
+        results = rouge.compute(predictions=[response], references=[[reference]], tokenizer=f, rouge_types=['rougeL'])
+        score = results['rougeL']
+        return score
+
+    def recall(
+        self,
+        nodes: list[NodeWithScore],
+        reference_doc: list[str]
+    ) -> float:
+        # 目前只考虑reference_doc 为 1 的情况
+
+        if nodes:
+            for node in nodes:
+                if reference_doc == node.text:
+                    return 1.0
+                else:
+                    continue
+        else:
+            return 0.0
+
+    async def SemanticSimilarity(
+            self,
+            response: str,
+            reference: str
+    ) -> float:
+
+        result = await self.evaluator.aevaluate(
+            response=response,
+            reference=reference,
+        )
+
+        return result.score
+
+    @staticmethod
+    def load_dataset():

```

Comment:
suggest here to be load_dataset(ds_names: list[str] = ["all"])  # ds_names=['dataset_name1', 'dataset_name2'], and return datasets